File: examples/bubble_sort.js
#!/usr/bin/env node

const arr = [6, 4, 0, 3, 2, 1, 6, 4, 0, 3, 2, 1, 6, 4, 0, 3, 2, 1, 6, 4, 0, 3, 2, 1, 6, 4, 0, 3, 2, 1, 6, 4, 0, 3, 2, 1];

async function sleep(ms) {
  return new Promise((resolve, _reject) => {
    setTimeout(resolve, ms)
  })
}

!async function() {
  for (let i = 0; i < arr.length - 1; i++) {
    for (let j = 0; j < arr.length - i; j++) {
      const left  = arr.at(j);
      const right = arr.at(j + 1);

      if (left > right) {
        console.log(`swap ${j} ${j + 1}`)
        arr[j]     = right;
        arr[j + 1] = left;
      }
      await sleep(1);
    }
  }
}()

File: examples/bubble_sort.rb
#!/usr/bin/env ruby

# This is the example provided in the README.md file

# Pssst, instead of doing `swap i j`, try the complex way:
# puts "set #{j} #{arr[j+1]}"
# puts "set #{j+1} #{arr[j]}"

# should be the same, right?

arr = [3, 9, 1, 8, 6, 10, 4, 7, 2, 5]

puts "[ #{arr.join(' ')} ]"

for i in 0..arr.length - 2 do
    for j in 0..arr.length - i - 2 do
        left = arr[j]
        right = arr[j + 1]

        # ⬇️ Here!
        puts "compare #{j} #{j + 1}"
        if left > right
            # ⬇️ And here!
            puts "swap #{j} #{j + 1}"
            arr[j], arr[j + 1] = right, left
        end
    end
end
